src/messenger/messenger.cpp

```diff
@@ -398,128 +404,225 @@ void Messenger::Terminate() {
+                               ZmqMessage::Build(msg_id, send_cb_id, remote_cb_id, router->identity_, message), true});
   }
 }
 
-common::ManagedPointer<ConnectionRouter> Messenger::GetConnectionRouter(const std::string &router_id) {
-  return common::ManagedPointer(routers_.at(router_id).get());
+callback_id_t Messenger::GetNextSendCallbackId() {
+  callback_id_t send_cb_id = next_callback_id_++;
+  // Check for wraparound.
+  if (0 == next_callback_id_.load().UnderlyingValue()) {
+    next_callback_id_.store(callback_id_t{static_cast<uint8_t>(BuiltinCallback::NUM_BUILTIN_CALLBACKS) + 1});
+    MESSENGER_LOG_INFO("[PID={}] Messenger: Message ID wrapped around!");
+  }
+  return send_cb_id;
 }
 
-common::ManagedPointer<CallbackFn> Messenger::GetCallback(uint64_t callback_id) {
-  return common::ManagedPointer(&callbacks_.at(callback_id));
+void Messenger::ServerLoopAddRouters() {
+  if (!routers_to_be_added_.empty()) {
+    std::lock_guard lock(routers_add_mutex_);
+    for (auto &item : routers_to_be_added_) {
+      auto router = std::make_unique<ConnectionRouter>(common::ManagedPointer(this), item.target_, item.identity_,
+                                                       item.callback_);
+      routers_.try_emplace(item.router_id_, std::move(router));
+    }
+    routers_to_be_added_.clear();
+    routers_add_cvar_.notify_all();
+  }
 }
 
-void Messenger::EraseCallback(uint64_t callback_id) { callbacks_.erase(callback_id); }
-
-uint64_t Messenger::GetNextSendMessageId() {
-  uint64_t send_msg_id = message_id_++;
-  // Check for wraparound.
-  if (0 == message_id_) {
-    message_id_ = static_cast<uint8_t>(BuiltinCallback::NUM_BUILTIN_CALLBACKS) + 1;
-    MESSENGER_LOG_INFO("[PID={}] Messenger: Message ID wrapped around!");
+void Messenger::ServerLoopMakeConnections() {
+  if (!connections_to_be_added_.empty()) {
+    std::lock_guard lock(connections_add_mutex_);
+    for (auto &item : connections_to_be_added_) {
+      auto connection = std::make_unique<ConnectionId>(common::ManagedPointer(this), item.target_, identity_);
+      connections_.try_emplace(item.connection_id_, std::move(connection));
+    }
+    connections_to_be_added_.clear();
+    connections_add_cvar_.notify_all();
   }
-  return send_msg_id;
 }
 
-void Messenger::ServerLoop() {
-  while (is_messenger_running_) {
-    // Add any new routers that need to be added.
-    // TODO(WAN): I wonder how expensive this block is.
-    if (!routers_to_be_added_.empty()) {
-      std::lock_guard lock(routers_add_mutex_);
-      for (auto &item : routers_to_be_added_) {
-        // TODO(WAN): I think I am missing something very obvious on how to handle this. The types involved suck.
-        auto router = std::make_unique<ConnectionRouter>(common::ManagedPointer(this), item.target_, item.identity_,
-                                                         item.callback_);
-        routers_.try_emplace(item.identity_, std::move(router));
+void Messenger::ServerLoopSendMessages() {
```

Comment:
Is it worth checking to see if `pending_messages_` is empty before acquiring the lock like the previous loop functions?

Reply:
Added. Note that this check-without-lock is technically undefined behavior. I added a comment as well.